packages/language-service/ivy/hybrid_visitor.ts

```diff
@@ -0,0 +1,169 @@
+        const {start, end} = isTemplateNode(last) ? getSpanIncludingEndTag(last) : last.sourceSpan;
+        const lastLength = end - start;
+        if (length > lastLength) {
+          // The current node has a span that is larger than the last node found
+          // so we do not descend into it. This typically means we have found
+          // a candidate in one of the root nodes so we do not need to visit
+          // other root nodes.
+          return;
+        }
+      }
+      this.path.push(node);
+      node.visit(this);
+    }
+  }
+
+  visitElement(element: t.Element) {
+    this.visitAll(element.attributes);
+    this.visitAll(element.inputs);
+    this.visitAll(element.outputs);
+    this.visitAll(element.references);
+    this.visitAll(element.children);
+  }
+
+  visitTemplate(template: t.Template) {
+    this.visitAll(template.attributes);
+    this.visitAll(template.inputs);
+    this.visitAll(template.outputs);
+    this.visitAll(template.templateAttrs);
+    this.visitAll(template.references);
+    this.visitAll(template.variables);
+    this.visitAll(template.children);
+  }
+
+  visitContent(content: t.Content) {
+    t.visitAll(this, content.attributes);
+  }
+
+  visitVariable(variable: t.Variable) {
+    // Variable has no template nodes or expression nodes.
+  }
+
+  visitReference(reference: t.Reference) {
+    // Reference has no template nodes or expression nodes.
+  }
+
+  visitTextAttribute(attribute: t.TextAttribute) {
+    // Text attribute has no template nodes or expression nodes.
+  }
+
+  visitBoundAttribute(attribute: t.BoundAttribute) {
+    const visitor = new ExpressionVisitor(this.position);
+    visitor.visit(attribute.value, this.path);
+  }
+
+  visitBoundEvent(attribute: t.BoundEvent) {
+    const visitor = new ExpressionVisitor(this.position);
+    visitor.visit(attribute.handler, this.path);
+  }
+
+  visitText(text: t.Text) {
+    // Text has no template nodes or expression nodes.
+  }
+
+  visitBoundText(text: t.BoundText) {
+    const visitor = new ExpressionVisitor(this.position);
+    visitor.visit(text.value, this.path);
+  }
+
+  visitIcu(icu: t.Icu) {
+    for (const boundText of Object.values(icu.vars)) {
+      this.visit(boundText);
+    }
+    for (const boundTextOrText of Object.values(icu.placeholders)) {
+      this.visit(boundTextOrText);
+    }
+  }
+
+  visitAll(nodes: t.Node[]) {
+    for (const node of nodes) {
+      this.visit(node);
+    }
+  }
+}
+
+class ExpressionVisitor extends e.RecursiveAstVisitor {
+  // Position must be absolute in the source file.
+  constructor(private readonly position: number) {
+    super();
+  }
+
+  visit(node: e.AST, path: Array<t.Node|e.AST>) {
+    if (node instanceof e.ASTWithSource) {
+      node = node.ast;
+    }
+    const {start, end} = node.sourceSpan;
+    // The third condition is to account for the implicit receiver, which should
+    // not be visited.
+    if (start <= this.position && this.position <= end && start < end) {
```

Comment:
what do you think of an explicit check for implicit receiver?

```suggestion
    // The third condition is to account for the implicit receiver, which should
    // not be visited.
    if (start <= this.position && this.position <= end && !(node instanceof e.ImplicitReceiver)) {
```